core/src/main/scala/kafka/controller/ControllerChannelManager.scala

```diff
@@ -397,24 +397,31 @@ class ControllerBrokerRequestBatch(controller: KafkaController, stateChangeLogge
       val stateChangeLog = stateChangeLogger.withControllerEpoch(controllerEpoch)
 
       val leaderAndIsrRequestVersion: Short =
-        if (controller.config.interBrokerProtocolVersion >= KAFKA_1_0_IV0) 1
+        if (controller.config.interBrokerProtocolVersion >= KAFKA_2_2_IV0) 2
+        else if (controller.config.interBrokerProtocolVersion >= KAFKA_1_0_IV0) 1
         else 0
 
       leaderAndIsrRequestMap.foreach { case (broker, leaderAndIsrPartitionStates) =>
-        leaderAndIsrPartitionStates.foreach { case (topicPartition, state) =>
-          val typeOfRequest =
-            if (broker == state.basePartitionState.leader) "become-leader"
-            else "become-follower"
-          stateChangeLog.trace(s"Sending $typeOfRequest LeaderAndIsr request $state to broker $broker for partition $topicPartition")
+        leaderAndIsrPartitionStates.foreach {
+          case (topicPartition, state) =>
+            val typeOfRequest =
+              if (broker == state.basePartitionState.leader) "become-leader"
+              else "become-follower"
+            stateChangeLog.trace(s"Sending $typeOfRequest LeaderAndIsr request $state to broker $broker for partition $topicPartition")
         }
-        val leaderIds = leaderAndIsrPartitionStates.map(_._2.basePartitionState.leader).toSet
-        val leaders = controllerContext.liveOrShuttingDownBrokers.filter(b => leaderIds.contains(b.id)).map {
-          _.node(controller.config.interBrokerListenerName)
+
+        controllerContext.brokerEpochs.get(broker) match {
+          case Some(brokerEpoch) =>
+            val leaderIds = leaderAndIsrPartitionStates.map(_._2.basePartitionState.leader).toSet
+            val leaders = controllerContext.liveOrShuttingDownBrokers.filter(b => leaderIds.contains(b.id)).map {
+              _.node(controller.config.interBrokerListenerName)
+            }
+            val leaderAndIsrRequestBuilder = new LeaderAndIsrRequest.Builder(leaderAndIsrRequestVersion, controllerId, controllerEpoch,
+              brokerEpoch, leaderAndIsrPartitionStates.asJava, leaders.asJava)
+            controller.sendRequest(broker, ApiKeys.LEADER_AND_ISR, leaderAndIsrRequestBuilder,
+              (r: AbstractResponse) => controller.eventManager.put(controller.LeaderAndIsrResponseReceived(r, broker)))
+          case None => stateChangeLog.trace(s"Broker epoch does not exists for broker $broker. Skip sending LeaderAndIsr request to this broker.")
```

Comment:
Thinking about this more, it may be better to do the following to explicit show that we want to keep only broker ids that are in `liveOrShuttingDownBrokerIds`.

```
leaderAndIsrRequestMap.filterKeys(controllerContext.liveOrShuttingDownBrokerIds.contains).foreach { case (broker, leaderAndIsrPartitionStates) =>
  ...
}
```

Reply:
Agree. Done.